The TransferLinksModule does not check if the link it tries to add to the parent is a link to the parent

## The Bug
It is impossible for obvious reasons to link a ticket to itself. When transferring links, Arisa does not check if the link it tries to add to a parent is a link to the parent, which results in Arisa attempting to add a self link to the parent.

```
10:57:25.436 [main] INFO  Arisa - [RESPONSE] [MC-180306] [TransferLinks] Successful
10:57:26.730 [main] ERROR Arisa - [UPDATE] [MC-150820] Failed
net.rcarz.jiraclient.JiraException: Failed to link issue MC-150820 with issue MC-150820
    at net.rcarz.jiraclient.Issue.link(Issue.java:1186)
    at net.rcarz.jiraclient.Issue.link(Issue.java:1124)
    at io.github.mojira.arisa.infrastructure.jira.FunctionsKt$createLink$1$1$1.invokeSuspend(Functions.kt:203)
    at io.github.mojira.arisa.infrastructure.jira.FunctionsKt$createLink$1$1$1.invoke(Functions.kt)
    at arrow.core.Either$Companion.catch(Either.kt:862)
    at arrow.core.Either$Companion.catch(Either.kt:858)
    at io.github.mojira.arisa.infrastructure.jira.FunctionsKt$createLink$1$1.invokeSuspend(Functions.kt:202)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
    at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:271)
    at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:79)
    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:54)
    at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:36)
    at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
    at io.github.mojira.arisa.infrastructure.jira.FunctionsKt$createLink$1.invoke(Functions.kt:201)
    at io.github.mojira.arisa.infrastructure.jira.FunctionsKt$createLink$1.invoke(Functions.kt)
    at io.github.mojira.arisa.modules.HelpersKt.tryRunAll(Helpers.kt:75)
    at io.github.mojira.arisa.infrastructure.jira.FunctionsKt.applyIssueChanges(Functions.kt:119)
    at io.github.mojira.arisa.ModuleExecutor.executeModule(ModuleExecutor.kt:109)
    at io.github.mojira.arisa.ModuleExecutor.execute(ModuleExecutor.kt:48)
    at io.github.mojira.arisa.ArisaMainKt.main(ArisaMain.kt:71)
    at io.github.mojira.arisa.ArisaMainKt.main(ArisaMain.kt)
Caused by: net.rcarz.jiraclient.RestException: 400 : {"errorMessages":["You cannot link an issue to itself."],"errors":{}}
    at net.rcarz.jiraclient.RestClient.request(RestClient.java:171)
    at net.rcarz.jiraclient.RestClient.request(RestClient.java:193)
    at net.rcarz.jiraclient.RestClient.request(RestClient.java:238)
    at net.rcarz.jiraclient.RestClient.post(RestClient.java:328)
    at net.rcarz.jiraclient.RestClient.post(RestClient.java:372)
    at net.rcarz.jiraclient.Issue.link(Issue.java:1184)
     ... 22 common frames omitted 
```

## Example Ticket
[MC-180306](https://bugs.mojang.com/browse/MC-180306)

## Expected behavior
Check if the link is a link to the parent being processed currently, and if so, don't add it.

--- a/src/main/kotlin/io/github/mojira/arisa/modules/TransferLinksModule.kt
+++ b/src/main/kotlin/io/github/mojira/arisa/modules/TransferLinksModule.kt
@@ -18,6 +18,7 @@ class TransferLinksModule : AbstractTransferFieldModule() {
             .flatMap { parent ->
                 links
                     .filter(::parentDoesNotHaveLink.partially1(parent.links))
+                    .filter(::doesNotLinkToParent.partially1(parent.key))
                     .map(::toLinkAdder.partially2(parent))
             }
 
@@ -30,6 +31,8 @@ class TransferLinksModule : AbstractTransferFieldModule() {
                 (it.type.toLowerCase() == "relates" || it.outwards == other.outwards) &&
                 it.issue.key == other.issue.key
         }
+
+    private fun doesNotLinkToParent(key: String, link: Link) = key != link.issue.key
 
     private fun toLinkAdder(
         link: Link,

--- a/src/test/kotlin/io/github/mojira/arisa/modules/TransferLinksModuleTest.kt
+++ b/src/test/kotlin/io/github/mojira/arisa/modules/TransferLinksModuleTest.kt
@@ -138,6 +138,41 @@ class TransferLinksModuleTest : StringSpec({
         hasTransferred.shouldBeFalse()
     }
 
+    "should not transfer links that link to the parent" {
+        var hasTransferred = false
+        val module = TransferLinksModule()
+
+        val createLink = { _: String, _: String, _: String -> hasTransferred = true; Unit.right() }
+
+        /**
+         * MC-42 duplicates MC-1
+         */
+        val duplicatesLink = linkIssues(
+            key1 = "MC-42",
+            key2 = "MC-1",
+            createLink = createLink
+        )
+
+        /**
+         * MC-42 relates to MC-1
+         */
+        val relatesLink = linkIssues(
+            key1 = "MC-42",
+            key2 = "MC-1",
+            type = "Relates"
+        )
+
+        val issue = mockIssue(
+            key = "MC-42",
+            links = listOf(duplicatesLink, relatesLink)
+        )
+
+        val result = module(issue, RIGHT_NOW)
+
+        result.shouldBeRight(ModuleResponse)
+        hasTransferred.shouldBeFalse()
+    }
+
     "should remove links" {
         var linkRemoved = false
         val module = TransferLinksModule()

